src/Geometry/CubeGenerator.cpp

```diff
@@ -0,0 +1,1582 @@
+    size_t start[2] = {x, 0};
+    size_t count[2] = {1, numElemPerPart[3]};
+    checkNcError(nc_put_vara_int(ncFile, ncVarElemGroup, start, count, elemGroup));
+  }
+  delete[] elemGroup;
+
+  // Vertices
+  std::map<int, CubeVertex> uniqueVertices;
+  transform(vertexMap.begin(),
+            vertexMap.end(),
+            inserter(uniqueVertices, uniqueVertices.begin()),
+            [](const auto& pair) { return std::make_pair(pair.second, pair.first); });
+  //&seissol::geometry::CubeGenerator::flip_pair<CubeVertex, int>);  //
+  //&<namespace>::<class>::<member_function> to pass pointer to member
+
+  int* vrtxSize = new int[numPartitions[3]];
+  std::fill(vrtxSize, vrtxSize + numPartitions[3], uniqueVertices.size());
+  checkNcError(nc_put_var_int(ncFile, ncVarVrtxSize, vrtxSize));
+  delete[] vrtxSize;
+  writes_done++;
+  loadBar(writes_done, netcdf_writes);
+
+  double* vrtxCoords = new double[uniqueVertices.size() * 3];
+
+  double halfWidthX = scaleX / 2.0;
+  double halfWidthY = scaleY / 2.0;
+  double halfWidthZ = scaleZ / 2.0;
+
+  for (unsigned int z = 0; z < numPartitions[2]; z++) {
+    for (unsigned int y = 0; y < numPartitions[1]; y++) {
+      for (unsigned int x = 0; x < numPartitions[0]; x++) {
+
+#pragma omp parallel for
+        for (unsigned int i = 0; i < uniqueVertices.size(); i++) {
+          vrtxCoords[i * 3] =
+              static_cast<double>(uniqueVertices.at(i).v[0] + x * numCubesPerPart[0]) /
+                  static_cast<double>(numCubes[0]) * scaleX -
+              halfWidthX + tx;
+          vrtxCoords[i * 3 + 1] =
+              static_cast<double>(uniqueVertices.at(i).v[1] + y * numCubesPerPart[1]) /
+                  static_cast<double>(numCubes[1]) * scaleY -
+              halfWidthY + ty;
+          vrtxCoords[i * 3 + 2] =
+              static_cast<double>(uniqueVertices.at(i).v[2] + z * numCubesPerPart[2]) /
+                  static_cast<double>(numCubes[2]) * scaleZ -
+              halfWidthZ + tz;
+        }
+
+        size_t start[3] = {(z * numPartitions[1] + y) * numPartitions[0] + x, 0, 0};
+        size_t count[3] = {1, uniqueVertices.size(), 3};
+        checkNcError(nc_put_vara_double(ncFile, ncVarVrtxCoords, start, count, vrtxCoords));
+        writes_done++;
+        loadBar(writes_done, netcdf_writes);
+      }
+    }
+  }
+
+  delete[] vrtxCoords;
+
+  checkNcError(nc_close(ncFile));
+
+  logInfo() << "Finished";
+}
+
+// Process had done i out of n rounds,
+// and we want a bar of width w and resolution r.
+// TODO put this i an library
+// default args from header file: r = 100, w = 50
+inline void seissol::geometry::CubeGenerator::loadBar(int x, int n, int r, int w) {
+  r = std::min(r, n);
+
+  // Only update r times.
+  if (x % (n / r) != 0)
+    return;
+
+  // Calculuate the ratio of complete-to-incomplete.
+  float ratio = x / (float)n;
+  int c = ratio * w;
+
+  // Show the percentage complete.
+  std::cout << std::setw(3) << (int)(ratio * 100) << "% [";
```

Comment:
Why `std::cout` here, and not e.g. `logInfo`?

(other than that, I like having a progress bar :) )

Reply:
That's a leftover from the original CubeGenerator code from `cube_c/`. I tried substituting the last `std::cout` call with `logInfo()` , but that results in a disjointed progress bar, e.g.
```
Thu Oct 05 15:20:53, Info:                              ]
Thu Oct 05 15:20:53, Info:  ===                         ]
Thu Oct 05 15:20:53, Info:  ========                    ]
Thu Oct 05 15:20:53, Info:  =============               ]
Thu Oct 05 15:20:53, Info:  ==================          ]
Thu Oct 05 15:20:53, Info:  =======================     ]
```
I could call `logInfo()` only when the progress bar is at 100% and write the progress into a `std::osstream`, but then for bigger files we would see no progress bar until the `CubeGenerator` is finished.
I can also try to create a logger object specifically for the progress bar to try to replicate the current behaviour, since AFAIK calling `logInfo()` always ends with a newline (\n).